crates/ast/src/attribute.rs

```diff
@@ -0,0 +1,78 @@
+use enum_map::{Enum, EnumMap};
+use enumflags2::{bitflags, BitFlags};
+use strum_macros::EnumString;
+
+/// An attribute that accepts a numeric value
+#[derive(Enum, Clone, Copy, PartialEq, EnumString)]
+pub enum NumAttr {
+    /// Tells the compiler to use a slow FSM design with a given II
+    #[strum(serialize = "slow_fsm")]
+    SlowFSM,
+}
+
+/// An flag attribute
+#[bitflags]
+#[repr(u8)]
+#[derive(Clone, Copy, PartialEq, EnumString)]
+pub enum BoolAttr {
+    /// This is a toplevel component
+    TopLevel,
+}
+
+/// Represents a single attribute. This is a private enum that is used during
+/// parsing to collect all attributes before creating the [Attributes] struct.
+pub enum Attr {
+    Bool(BoolAttr),
+    Num(NumAttr, u64),
+}
+
+/// A set of attributes attached to a component
+#[derive(Default, Clone)]
+pub struct Attributes {
+    bool_attrs: BitFlags<BoolAttr>,
+    num_attrs: EnumMap<NumAttr, Option<u64>>,
```

Comment:
Is there a reason this can be `Option<u64>`? Wouldn't that make it a `BoolAttr`?

Reply:
Would this not limit numeric attributes to 0 as a default value? I was thinking of adding passes that set the values if they dont exist (see FSMAttributes pass)

Reply:
Hmm, I was thinking that if the `Attr` was not defined, it would not be present in the map (this is how Calyx attributes work).

Reply:
I think the map is represented as an array under the hood so everything is initialized to default (0)...

Reply:
I chose to use this instead of the calyx linkedhashmap due to performance, but I could do it that way instead?

Reply:
Okay, got it. `Option<T>` is bigger than just `T` which is why I was wondering. Either way, doesn't matter too much I suspect.